### PR TITLE
Improve the storage method of the grant info in the metastore

### PR DESCRIPTION
**Pull request by SimFG**
issue: #18816 
More details about the improvement way can be seen above issue link.

Signed-off-by: SimFG <bang.fu@zilliz.com>

### CODE REVIEW

File: internal/metastore/kv/rootcoord/kv_catalog.go [hunk 1]
```diff
@@ -1009,85 +1015,76 @@ func (kc *Catalog) AlterGrant(ctx context.Context, tenant string, entity *milvus
...
 			return err
 		}
-		curGrantPrivilegeEntity.Entities = append(curGrantPrivilegeEntity.Entities, &milvuspb.GrantorEntity{
-			Privilege: &milvuspb.PrivilegeEntity{Name: privilegeName},
-			User:      &milvuspb.UserEntity{Name: entity.Grantor.User.Name},
-		})
-	} else {
-		err = proto.Unmarshal([]byte(v), curGrantPrivilegeEntity)
+		var id typeutil.UniqueID
+		id, _, err = kc.idAllocate(1)
```
Comment (jaime0815) [inline internal/metastore/kv/rootcoord/kv_catalog.go:1019]: Suggest removing idAllocate method and using a hash value of the key to replacing it.

**Comment by SimFG** [comment]
rerun ut

**Comment by SimFG** [comment]
rerun ut

File: internal/metastore/kv/rootcoord/kv_catalog.go [hunk 2]
```diff
@@ -880,7 +873,7 @@ func (kc *Catalog) ListGrant(ctx context.Context, tenant string, entity *milvusp
 		for i, key := range keys {
 			grantInfos := typeutil.AfterN(key, k+"/", "/")
 			if len(grantInfos) != 2 {
-				log.Warn("invalid grant key", zap.String("key", key))
+				log.Warn("invalid grant key", zap.String("key", key), zap.String("k", k))
```
Comment (congqixia) [inline internal/metastore/kv/rootcoord/kv_catalog.go:876]: key, k, grantKey is kind of confusing here

File: internal/metastore/db/dao/gran_id.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	err := g.db.Create(in).Error
+	if err != nil {
+		log.Error("fail to insert the grant-id", zap.Any("in", in), zap.Error(err))
+	}
+	return err
+}
+
+func (g *grantIDDb) Delete(tenantID string, grantID int64, privilege string) error {
+	err := g.db.Model(dbmodel.GrantID{}).
+		Where(&dbmodel.GrantID{GrantID: grantID, Privilege: privilege}).
```
[the remark this answers is not in the document]
Comment (SimFG) [reply internal/metastore/db/dao/gran_id.go:47]: This part is to provide a general capability, and the specific usage scenario is determined by the upper layer.

File: internal/metastore/kv/rootcoord/kv_catalog.go (hunk 2 above)
Comment (SimFG) [reply internal/metastore/kv/rootcoord/kv_catalog.go:876]: Updated.

File: internal/metastore/db/dao/grant.go [hunk 4]
```diff
@@ -30,64 +29,20 @@ func (g *grantDb) GetGrants(tenantID string, roleID int64, object string, object
...
-	}
-	db = g.db.Model(dbmodel.Grant{}).Where(sqlWhere).Where(dbmodel.GetCommonCondition(tenantID, false))
-	if newDbDetail == "" {
-		err = db.Update("is_deleted", true).Error
-	} else {
-		err = db.Update("detail", newDbDetail).Error
-	}
+func (g *grantDb) Delete(tenantID string, roleID int64, object string, objectName string) error {
+	err := g.db.Model(dbmodel.Grant{}).
+		Where(&dbmodel.Grant{RoleID: roleID, Object: object, ObjectName: objectName}).
```
[the remark this answers is not in the document]
Comment (congqixia) [reply internal/metastore/db/dao/grant.go:41]: Still using struct as condition here.

**Review by congqixia — COMMENTED** [review]
And I just realized all db operation has no context parameter

**Comment by SimFG** [comment]
Because the `metastore`  is designed that the `context` param isn't included. If needed later, it will be added.

And the params of the `delete` and `get` functions is the condition, and I think using the struct is unnecessary. If doing it, using the model object isn't suitable and defining a new struct is complicated.
@congqixia

**Review by congqixia — APPROVED** [review]
NIT: maybe add some comment for delete parameter, especially the zero value case
/lgtm

**Pull request merged**